javarosa-dag.rst

```diff
@@ -0,0 +1,304 @@
+    <bind nodeset="/data/nodeB" calculate="pow(/data/nodeC, 2)" type="int" />
+    <bind nodeset="/data/nodeC" type="int"/>
+
+In the above snippet, both ``nodeA`` and ``nodeB`` have the same ``calculation`` expression which is evaluated
+when ``nodeC`` is updated. The ``nodeC`` is the *trigger* of ``nodeA`` and ``nodeB`` calculations.
+
+To avoid storing two separate but equal``Triggerable`` instances, the ``IDag.addTriggerable(Triggerable t)``
+method first checks if a same triggerable has been already added. 
+Two ``Triggerable`` instances are considered equal if:
+
+* Both have the same implementation type - ``Recalculate`` never equals ``Condition``.
+* Both have the same expressions to be evaluated.
+* Both have the same triggers. This is implied by the second bullet as triggers are the nodes referenced in the expressions.
+
+If such triggerable already exists, ``IDag`` does not add a new instance but updates the existing``Triggerable`` one's context so it points to the 
+highest common root. For the above example, the highest common root for ``nodeA`` and ``nodeB`` calculations is ``/data``.
+The passed instance ``t`` is ignored and a previously added ``Triggerable`` one is returned and should be used for the further processing.
+The context is used later when constructing a DAG structure.
+
+If such triggerable does not exist, it is added to the ``unorderedTriggerables`` list and the ``triggerIndex`` map.
+The passed triggerable ``t`` instance is returned.
+The list is later used to construct the topoligically sorted list ``triggerablesDAG``.
+The ``triggerIndex`` maps the triggers to the triggerables.
+
+Each ``Triggerable`` instance comes with a list of targets which should be updated with the evaluation result.
+In the above example, ``pow(/data/nodeC, 2)`` calculation have two targets: ``nodeA`` and ``nodeB``.
+When a ``Triggerable`` instance is constructed, the targets list cannot be populated because each ``bind`` node
+is parsed separately and other targets - if any - for the same triggerable are unknown. Moreover,
+the newly constructed ``Triggerable`` may be ignored if another instance that represents the same triggerable
+have been already added to the DAG.
+
+.. Actually, the above is not entirely true. I think it can be refactored so the targets are populated during 
+.. the bind processing. But a particular attention must be paid to call ``IDag.addTriggerable(Triggerable t)`` before
+.. ``Triggerable.addTarget(TreeReference target)`` because  ``IDag.addTriggerable(Triggerable t)` may return a previously added
+.. instance and there fore ``t`` is ignored.
+
+The ``targets`` list population is postponed and takes place during the form instance parsing, 
+in the ``FormInstanceParser.applyInstanceProperties(FormInstance instance)`` method.
+
+Finalization and constructing the DAG
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Once all ``bind`` nodes are parsed and all ``Triggerable`` collected, the ``IDag`` can be finalized and
```

Comment:
and all ``Triggerable``s are collected

Reply:
Won't work. Will require a space between `Triggerable` and 's'

Reply:
How about “Triggerable objects” or something like that, then?

Reply:
@lognaturel Sounds good, will do!